File: refinery-service-java/src/main/java/hu/zoltanterek/worldsim/refinery/contracts/RefineryVocabulary.java
package hu.zoltanterek.worldsim.refinery.contracts;

import java.util.List;
import java.util.Set;

public final class RefineryVocabulary {
    private RefineryVocabulary() {
    }

    public static final String OUTPUT_MODE_BOTH = "both";
    public static final String OUTPUT_MODE_STORY_ONLY = "story_only";
    public static final String OUTPUT_MODE_NUDGE_ONLY = "nudge_only";
    public static final String OUTPUT_MODE_OFF = "off";

    public static final String SEVERITY_MINOR = "minor";
    public static final String SEVERITY_MAJOR = "major";
    public static final String SEVERITY_EPIC = "epic";

    public static final String EFFECT_TYPE_DOMAIN_MODIFIER = "domain_modifier";
    public static final String BIAS_TYPE_GOAL_BIAS = "goal_bias";

    public static final String CAMPAIGN_KIND_DECLARE_WAR = "declare_war";
    public static final String CAMPAIGN_KIND_PROPOSE_TREATY = "propose_treaty";
    public static final String TREATY_KIND_CEASEFIRE = "ceasefire";
    public static final String TREATY_KIND_PEACE_TALKS = "peace_talks";

    public static final List<String> SHARED_OUTPUT_MODES = List.of(
            OUTPUT_MODE_BOTH,
            OUTPUT_MODE_STORY_ONLY,
            OUTPUT_MODE_NUDGE_ONLY,
            OUTPUT_MODE_OFF
    );

    public static final Set<String> SEVERITIES = Set.of(SEVERITY_MINOR, SEVERITY_MAJOR, SEVERITY_EPIC);

    public static final Set<String> DOMAINS = Set.of(
            "food",
            "morale",
            "economy",
            "military",
            "research"
    );

    public static final Set<String> GOAL_CATEGORIES = Set.of(
            "farming",
            "gathering",
            "crafting",
            "building",
            "social",
            "military",
            "research",
            "rest"
    );

    public static final Set<String> CAMPAIGN_KINDS = Set.of(CAMPAIGN_KIND_DECLARE_WAR, CAMPAIGN_KIND_PROPOSE_TREATY);
    public static final Set<String> TREATY_KINDS = Set.of(TREATY_KIND_CEASEFIRE, TREATY_KIND_PEACE_TALKS);
}

File: WorldSim.Contracts/v2/RefineryVocabulary.cs
namespace WorldSim.Contracts.V2;

public static class RefineryVocabulary
{
    public const string OutputModeBoth = "both";
    public const string OutputModeStoryOnly = "story_only";
    public const string OutputModeNudgeOnly = "nudge_only";
    public const string OutputModeOff = "off";

    public const string SeverityMinor = "minor";
    public const string SeverityMajor = "major";
    public const string SeverityEpic = "epic";

    public const string EffectTypeDomainModifier = "domain_modifier";
    public const string BiasTypeGoalBias = "goal_bias";

    public const string TreatyKindCeasefire = "ceasefire";
    public const string TreatyKindPeaceTalks = "peace_talks";

    public static readonly IReadOnlyList<string> SharedOutputModes =
    [
        OutputModeBoth,
        OutputModeStoryOnly,
        OutputModeNudgeOnly,
        OutputModeOff
    ];

    public static readonly IReadOnlyList<string> Severities =
    [
        SeverityMinor,
        SeverityMajor,
        SeverityEpic
    ];

    public static readonly IReadOnlyList<string> Domains =
    [
        "food",
        "morale",
        "economy",
        "military",
        "research"
    ];

    public static readonly IReadOnlyList<string> GoalCategories =
    [
        "farming",
        "gathering",
        "crafting",
        "building",
        "social",
        "military",
        "research",
        "rest"
    ];

    public static readonly IReadOnlyList<string> TreatyKinds =
    [
        TreatyKindCeasefire,
        TreatyKindPeaceTalks
    ];

    public static bool IsSharedOutputMode(string mode) => SharedOutputModes.Contains(mode, StringComparer.Ordinal);

    public static bool IsTreatyKind(string treatyKind) => TreatyKinds.Contains(treatyKind, StringComparer.Ordinal);

    public static bool IsSeverity(string severity) => Severities.Contains(severity, StringComparer.Ordinal);
}
